sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -121,6 +129,24 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
         if canImplicitlyCast(fromExp, toType, literalType) =>
       simplifyNumericComparison(be, fromExp, toType, value)
 
+    case in @ In(Cast(fromExp, toType: NumericType, _), list)
+      if list.forall(v =>
+        v.isInstanceOf[Literal] && canImplicitlyCast(fromExp, toType, v.dataType)
+      ) =>
+      val (newValueList, exp) =
+        list.map(lit => unwrapCast(EqualTo(in.value, lit)))
+          .partition {
```

Comment:
ditto, the list is always of the same data type. We don't need to do `partition` here.

Reply:
As @sunchao  mentioned before, we `partition` results because the `unwrapCast` converts a `EqualTo` to a different expression. 

For instance:

```
x = 1 => x = 1
```
```
x = 2147483648 => if(isnull(x), null, false)
```

if `x` is of integer type.

Do you mean we can simply do `x in (1, 2147483648) => x in (1)`?

Reply:
ah I see, I missed that.